pkg/storage/rulestore/config.go

```diff
@@ -0,0 +1,122 @@
+package rulestore
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/objstore"
+
+	"github.com/cortexproject/cortex/pkg/configs/client"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/azure"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/filesystem"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/gcs"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/s3"
+	"github.com/cortexproject/cortex/pkg/storage/bucket/swift"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	ConfigDB = "configdb"
+
+	name   = "rule-store"
+	prefix = "rule-store."
+)
+
+var (
+	supportedBackends = []string{bucket.S3, bucket.GCS, bucket.Azure, bucket.Swift, bucket.Filesystem, ConfigDB}
+
+	ErrUnsupportedStorageBackend = errors.New("unsupported storage backend")
+)
+
+// Config configures a rule store.
+type Config struct {
+	bucket.Config `yaml:",inline"`
+	ConfigDB      client.Config `yaml:"configdb"`
+}
+
+// RegisterFlags registers the backend storage config.
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.ConfigDB.RegisterFlagsWithPrefix(prefix, f)
+
+	cfg.S3.RegisterFlagsWithPrefix(prefix, f)
+	cfg.GCS.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Azure.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Swift.RegisterFlagsWithPrefix(prefix, f)
+	cfg.Filesystem.RegisterFlagsWithPrefix(prefix, f)
+
+	f.StringVar(&cfg.Backend, prefix+"backend", "s3", fmt.Sprintf("Backend storage to use. Supported backends are: %s.", strings.Join(supportedBackends, ", ")))
+}
+
+func (cfg *Config) Validate() error {
+	if !util.StringsContain(supportedBackends, cfg.Backend) {
+		return ErrUnsupportedStorageBackend
+	}
+
+	if cfg.Backend == bucket.S3 {
+		if err := cfg.S3.Validate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NewRuleStore creates a new bucket client based on the configured backend
+func NewRuleStore(ctx context.Context, cfg Config, logger log.Logger, reg prometheus.Registerer) (rules.RuleStore, error) {
```

Comment:
Why this function contains a bunch of duplicated code from `bucket.NewClient()`? We should reuse it.

Also configdb is not supported yet.

Reply:
Makes sense, I'll refactor to reuse upstream and implement the configdb.